File: fix_stream_builders.dart
import 'dart:io';

void main() async {
  final libDir = Directory('lib');
  if (!await libDir.exists()) {
    print('lib directory not found');
    return;
  }

  const importStatement = "import 'package:edukn/widgets/safe_stream_builder.dart';\n";
  int filesModified = 0;

  final files = libDir.listSync(recursive: true);
  for (final entity in files) {
    if (entity is File && entity.path.endsWith('.dart') && !entity.path.contains('safe_stream_builder.dart')) {
      String content;
      try {
        content = await entity.readAsString();
      } catch (e) {
        try {
           content = await entity.readAsString(encoding: SystemEncoding());
        } catch (e2) {
           print('Skipping ${entity.path} due to encoding error');
           continue;
        }
      }
      
      // Look for StreamBuilder< or StreamBuilder(
      final streamBuilderRegex = RegExp(r'(?<!Safe)StreamBuilder[\s]*[<\(]');
      
      if (streamBuilderRegex.hasMatch(content)) {
        String newContent = content.replaceAllMapped(streamBuilderRegex, (match) {
          final suffix = match.group(0)!.substring(match.group(0)!.length - 1);
          return 'SafeStreamBuilder$suffix';
        });

        if (!newContent.contains("import 'package:edukn/widgets/safe_stream_builder.dart';")) {
          final importRegex = RegExp(r'^import\s+.*;$', multiLine: true);
          final matches = importRegex.allMatches(newContent);
          
          if (matches.isNotEmpty) {
            final lastMatch = matches.last;
            final insertPos = lastMatch.end + 1;
            newContent = newContent.substring(0, insertPos) + importStatement + newContent.substring(insertPos);
          } else {
            newContent = importStatement + newContent;
          }
        }
        
        await entity.writeAsString(newContent);
        filesModified++;
        print('Updated: ${entity.path}');
      }
    }
  }
  
  print('Total files updated: $filesModified');
}

File: fix_stream_builders.py
import os
import re

lib_dir = r"c:\Users\user\Desktop\eduKN\edukn\edukn21.11.2025\edukn\lib"
import_statement = "import 'package:edukn/widgets/safe_stream_builder.dart';\n"
files_modified = 0

for root, _, files in os.walk(lib_dir):
    for file in files:
        if file.endswith('.dart') and file != 'safe_stream_builder.dart':
            filepath = os.path.join(root, file)
            
            with open(filepath, 'r', encoding='utf-8') as f:
                content = f.read()
                
            # Check if StreamBuilder is used
            # We look for "StreamBuilder<" or "StreamBuilder("
            # using regex that makes sure it's not preceded by "Safe"
            if re.search(r'(?<!Safe)StreamBuilder[\s]*[<\(]', content):
                # Replace it
                new_content = re.sub(r'(?<!Safe)StreamBuilder([\s]*[<\(])', r'SafeStreamBuilder\1', content)
                
                # Add import if not present
                if "import 'package:edukn/widgets/safe_stream_builder.dart';" not in new_content:
                    # Find the last import
                    import_matches = list(re.finditer(r'^import\s+.*;$', new_content, re.MULTILINE))
                    if import_matches:
                        last_import = import_matches[-1]
                        insert_pos = last_import.end() + 1
                        new_content = new_content[:insert_pos] + import_statement + new_content[insert_pos:]
                    else:
                        new_content = import_statement + new_content
                
                with open(filepath, 'w', encoding='utf-8') as f:
                    f.write(new_content)
                files_modified += 1
                print(f"Updated: {filepath}")

print(f"Total files updated: {files_modified}")
